js/editor-controller.js

```diff
@@ -0,0 +1,235 @@
+                            throw new Error("'projects' array should contain at least one item.");
+                        }
+                        _editorActions.setCode(event.data.projects[0]);
+                        // Notify parent about editor successfully configured
+                        window.parent.postMessage({
+                            type: CONTROLLER_MESSAGING.type,
+                            action: CONTROLLER_MESSAGING.actions.workspaceloaded
+                        }, '*');
+                        break;
+
+                    // Parent is sending HEX file
+                    case CONTROLLER_MESSAGING.actions.loadhex:
+                        _editorActions.loadHex(event.data.filename, event.data.hexstring);
+                        break;
+
+                    // Parent is sending file for filesystem
+                    case CONTROLLER_MESSAGING.actions.loadfile:
+                        _editorActions.loadFileToFs(event.data.filename, event.data.filestring);
+                        break;
+
+                    // Parent is requesting postMessage downloads
+                    case CONTROLLER_MESSAGING.actions.downloadmode:
+                        _editorActions.setMobileEditor(hostFlashHex, hostSaveFile);
+                        break;
+
+                    default:
+                        throw new Error('Unsupported action.')
+                }
+            }
+        }, false);
+    };
+
+    /**
+     * Initialises the controller based on the URL query string parameters and
+     * the `window` environment.
+     * The Classroom controller will be an iframe parent, and the mobile apps
+     * will have postMessage handlers on different namespaces within `window`.
+     *
+     * @param {object} qs - The editor URL query string parameters.
+     */
+    _returnObj.initialise = function(qs) {
+        var inIframe = window !== window.parent;
+        var iframeControllerMode = inIframe && qs.controller == '1';
+        var appControllerMode = qs.mobileApp == '1';
+
+        if (iframeControllerMode || appControllerMode) {
+            // Detect the host controller to send our messages
+            if (iframeControllerMode && window.parent && window.parent.postMessage) {
+                // Classroom wraps the editor in an iframe
+                _controllerHost = window.parent;
+            } else if (appControllerMode &&
+                    window.webkit &&
+                    window.webkit.messageHandlers &&
+                    window.webkit.messageHandlers.host &&
+                    window.webkit.messageHandlers.host.postMessage) {
+                // iOS app, 'host' used to match the MakeCode messageHandler name
+                _controllerHost = window.webkit.messageHandlers.host;
+            } else if (appControllerMode && window.host && window.host.postMessage) {
+                // Android app
+                _controllerHost = window.host
```

Comment:
@microbit-sam is window.host defined by the Android webview, or is it a namespace created by us?

Reply:
Yep - Android creates the name when setting up the native handler

`webView.addJavascriptInterface(new JsObject(), "host");`
`window.host.postMessage(msg)`

Reply:
Right, so the namespace "host" is defined by us in the app code.

Reply:
Yeah we define it, as long as it's the same in the app and python editor it will work